qcodes/instrument/parameter.py

```diff
@@ -1387,9 +1473,14 @@ def snapshot_base(self, update: Optional[bool] = True,
             update=update,
             params_to_skip_update=params_to_skip_update
         )
-        snapshot.update(
-            {'source_parameter': self.source.snapshot(update=update)}
-        )
+        if self.source is None:
+            snapshot.update(
+                {'source_parameter': None}
+            )
+        else:
+            snapshot.update(
+                {'source_parameter': self.source.snapshot(update=update)}
+            )
```

Comment:
perhaps smth like this in order not to duplicate the "source_parameter" key?
```suggestion
        source_parameter_snapshot = None if self.source is None \
            else self.source.snapshot(update=update)
        snapshot.update(
            {'source_parameter': source_parameter_snapshot}
        )
```

Reply:
Not crazy about the line break but sure